kart/spatial_tree.py

```diff
@@ -0,0 +1,413 @@
+        if transforms is None:
+            transforms = self._load_transforms_for_dataset(ds_path)
+            self.ds_to_transforms[ds_path] = transforms
+        return transforms
+
+    def _load_transforms_for_dataset(self, ds_path):
+        if ds_path in self.ds_to_transforms:
+            return self.ds_to_transforms[ds_path]
+
+        crs_oids = set(self.iter_crs_oids(ds_path))
+        transforms = []
+        descs = []
+        for crs_oid in crs_oids:
+            try:
+                transform, desc = self.transform_from_oid(crs_oid)
+                if transform not in transforms:
+                    transforms.append(transform)
+                    descs.append(desc)
+            except Exception as e:
+                L.warning(
+                    f"Couldn't load transform for CRS {crs_oid} at {ds_path}\n{e}"
+                )
+        L.info(f"Loaded CRS transforms for {ds_path}: {', '.join(descs)}")
+        return transforms
+
+    def iter_crs_oids(self, ds_path):
+        cmd = (
+            _revlist_command(self.repo)
+            + ["--all", "--"]
+            + list(self.all_crs_paths(ds_path))
+        )
+        try:
+            r = subprocess.run(
+                cmd,
+                encoding="utf8",
+                check=True,
+                capture_output=True,
+                env=tool_environment(),
+            )
+        except subprocess.CalledProcessError as e:
+            raise SubprocessError(
+                f"There was a problem with git rev-list: {e}", called_process_error=e
+            )
+        for d, crs_oid in _parse_revlist_output(
+            r.stdout.splitlines(), r"meta/crs/[^/]+"
+        ):
+            assert d == ds_path
+            yield crs_oid
+
+    def all_crs_paths(self, ds_path):
+        # Delete .sno-dataset if we drop V2 support.
+        yield f"{ds_path}/.sno-dataset/meta/crs/"
+        yield f"{ds_path}/.table-dataset/meta/crs/"
+
+    @functools.lru_cache()
+    def transform_from_oid(self, crs_oid):
+        wkt = normalise_wkt(self.repo[crs_oid].data.decode("utf-8"))
+        return self.transform_from_wkt(wkt)
+
+    @functools.lru_cache()
+    def transform_from_wkt(self, wkt):
+        src_crs = make_crs(wkt)
+        if src_crs.IsGeographic():
+            transform = None
+            desc = f"IDENTITY({src_crs.GetAuthorityCode(None)})"
+        else:
+            target_crs = src_crs.CloneGeogCS()
+            transform = osr.CoordinateTransformation(src_crs, target_crs)
+            desc = f"{src_crs.GetAuthorityCode(None)} -> {target_crs.GetAuthorityCode(None)}"
+        return transform, desc
+
+
+class SpatialTreeTables(TableSet):
+    """Spatial Tree spec tables - see http://www.geopackage.org/spec/#table_definition_sql"""
+
+    def __init__(self):
+        super().__init__()
+
+        self.blobs = Table(
+            "blobs",
+            self.sqlalchemy_metadata,
+            Column("rowid", Integer, nullable=False, primary_key=True),
```

Comment:
can you add a comment to each of these columns describing what they store? With examples if possible

Reply:
Done